src/main/java/org/kiwiproject/consul/util/failover/strategy/BlacklistingConsulFailoverStrategy.java

```diff
@@ -91,43 +82,75 @@ public Optional<Request> computeNextStage(@NonNull Request previousRequest, @Nul
      */
     private Optional<HostAndPort> findTargetNotInBlacklist() {
         return targets.stream().filter(target -> {
+            if (isNotBlacklisted(target)) {
+                return true;
+            }
 
-            // If we have blacklisted this key
-            if (blacklist.containsKey(target)) {
-
-                // Get when we blacklisted this key
-                final Instant blacklistedAt = blacklist.get(target);
-
-                // If the duration between the blacklist time ("then") and "now" is greater than the
-                // timeout duration (Duration(then, now) - timeout < 0), then remove the blacklist entry
-                if (Duration.between(blacklistedAt, Instant.now()).minusMillis(timeout).isNegative()) {
-                    return false;
-                } else {
-                    blacklist.remove(target);
-                    return true;
-                }
-            } else {
+            if (isPastBlacklistDuration(target)) {
+                // Remove the target from blacklist once timeout has expired
+                blacklist.remove(target);
                 return true;
             }
+
+            return false;
         }).findAny();
     }
 
+    @VisibleForTesting
+    boolean isPastBlacklistDuration(HostAndPort target) {
+        var blacklistedAt = blacklist.get(target);
+
+        // Exit early if the target is not actually in the blacklist.
+        // The only way this can be true is if a concurrent request causes the
+        // target to be removed from the blacklist between the time it is first
+        // checked in #findTargetNotInBlacklist and when this method gets the
+        // value of the target key above. If that happens, just return true. It
+        // won't cause any problems with the Map#remove operation since that
+        // will just return null if there was no mapping with the given key.
+        if (isNull(blacklistedAt)) {
+            return true;
+        }
+
+        // If the duration between the blacklist time ("then") and "now" is greater than the
+        // timeout duration (Duration(then, now) - timeout >= 0), then the timeout has passed
+        var adjustedBlacklistDuration = Duration.between(blacklistedAt, Instant.now()).minusMillis(timeout);
```

Comment:
could be simpler to use before/after
```
return Instant.now().isAfter(blacklistedAt.addMillis(timeout))
```
though it would change the logic slightly since the expiration happens `after` the timeout instead of `at or after`

Reply:
Yes, I agree that is more understandable.

It's a shame `Instant` does not have `isAfterOrEqualTo`, but to keep the logic the same, how about:

```java
// If "now" is after or equal to the time target was blacklisted plus the timeout amount,
// then the timeout has passed
var timeoutExpiration = blacklistedAt.plusMillis(timeout);
var now = Instant.now();
return now.equals(timeoutExpiration) || now.isAfter(timeoutExpiration);
```